llvm/lib/CodeGen/GlobalISel/CallLowering.cpp

```diff
@@ -790,6 +804,36 @@ bool CallLowering::handleAssignments(ValueHandler &Handler,
       CCValAssign &VA = ArgLocs[j + Idx];
       const ISD::ArgFlagsTy Flags = Args[i].Flags[Part];
 
+      // We found an indirect parameter passing and we are at the first part of
+      // the value being passed. In this case copy the incoming pointer into a
+      // virtual register so later we can load it.
+      if (VA.getLocInfo() == CCValAssign::Indirect && Flags.isSplit()) {
```

Comment:
I think this should be done for unsplit arguments as well.

Reply:
Only the first part is processed, when we have indirect passing. I wrote about it in the previous comment.

This is how things happen:
1. `RISCVCallLowering::lowerFormalArguments` is called.
2. The arguments are prepared, ArgInfor array is constructed in a loop.
3. `CallLowering::determineAssignments` is called.
4. In `CallLowering::determineAssignments` we decide how many registers would be needed to store the value being passed.
5. Then we create as many  `CCValAssign`s as many register would be needed.
6. We run the calling convention function on all of them (CallLowering.cpp:681 the loop). This loop marks the first register with setSplit and the last with setSplitEnd
7. In the calling convention function, we recognize that we need indirect passing(The RISCV::CC_RISCV knows when we need that. This code is also used by SelDAG). It sets the `CCValAssign` to be an indirect assignment.
8. After this we get back to handle assignment, where in the `CCState` we have the informations to handle the assignments.
9. And this is where my PR comes in and handles the assignments.

This is my understanding.

Reply:
So basically checking `isSplit` is just a safety check. Maybe it should be also enough info, that we have an indirectly passed argument and we are in the first "part" of it, so we should handle the indirect passing.

Reply:
In p4 we can decide that one register is enough, in which case isSplit will be false.
This is probably not applicable to RISC-V, but this is a generic code and should support such scenario.
If this is not going to be supported in this patch, I'd expect a fatal error if isSplit is false when Indirect is set.

Reply:
> I wanted to disable indirect parameter passing for vectors, but I forgot to do that. Or should we do this indirect passing for vector types too?

If a target decided to pass an argument indirectly (via determineAssignments), we must do so regardless of the argument's type.